Export selected cards from CardBrowser

TarekkMA opened the pull request:
Export individual cards from CardBrowser

## Pull Request template

## Purpose / Description

> Following upstream, I believe it would be nice to be able to select card(s) in the deck browser and have an option to export them. It would be especially useful to share buggy card for debugging or to share a card type with an example without needing to share a whole deck or create it



## Fixes
Fixes #8045 

<img src="https://user-images.githubusercontent.com/6633545/110561577-9a3d2f80-8150-11eb-9415-160136c31cbe.png" width="25%"><img src="https://user-images.githubusercontent.com/6633545/110561582-9c06f300-8150-11eb-825c-1c79d09947f0.png" width="25%">

## Approach
I have extracted all the needed logic for exporting cards and moved it to its own class ``ExportHelper``. This class will delegate all actions from any activity that wish to implement exporting.

I also needed to change the ankilib to make it a colse as the pylib one

## How Has This Been Tested?

- [x] Export normal deck
- [x] Export collection
- [x] Export individual cards from card browser

## Learning (optional, can help others)
_Describe the research stage_

_Links to blog posts, patterns, libraries or addons used to solve this problem_

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)

Arthur-Milchior reviewed (commented):
Hi,

First of all, thank's very much for this PR, I REALLY appreciate it. It's a feature I wanted and that's really really helpful when people have bug on cards. This way, they can export the card with the bug and only it, and it's easier to debug. (Thinking about it, adding this export feature on the reviewer/card editor maybe nice too. It seems easier to export a buggy card when you see the bug than to look for it in the browser. This will need to be considered in a separate feature request)

I also love the fact that it uses more background interfacing. I rewrote this previously, and I'm happy to see someone else using this, showing that it's actually usable.
If you please, I'd love you to take a look at AnkiDroid/src/test/java/com/ichi2/libanki/AnkiPackageExporterTest.java and see if you can add test too. That would be wonderful.

I really appreciate that you take time cleaning some code, adding comments, renaming a file in order to be accurate, splitting logically a part of a huge class in a smaller one. That's really a great first PR.

I have one request however, that I would really appreciate that you would consider. This first commit is really hard to read and validate because it does a lot of things simultaneously. Some things are meaningful and add the feature we want, and that's great. Some other things are related to cleaning, moving code, and that's great too, but that requires less attention.
So if you could do multiple commits, such as:
* One where you just rename the file
* One where you do the code cleaning you wanted to do
* One where you move a part of DeckPicker to a new class
* And finally one where you do the actual change that add the new features
that would allow mo to quickly check all simple commit and then concentrate on the important one. Currently, it's hard to be sure I didn't miss anything

If you are familiar with git, hopefully, this should not be painful to do. Otherwise, don't hesitate to ask for help and I'll be happy to help you learn how to split a commit in smaller atomic commits.


You contacted me in PM on discord. I believe this commit is related to GSoC, is it right?
You seems to have already a quite good level on Android programming. That's not exactly a big change, but that used UI, string, background/foreground communication, adding abstraction were needed. That's really a cool PR, but I'm also worried that this internship, if you were selected, would be a little bit boring. Clearly, you can do great contributions, but I do not know on our side how much we would be able to teach you, which is also an important part of internship. 
Of course, it depends on what you want to work on. But I felt I should mention that.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:
```diff
@@ -349,6 +375,45 @@ public void actualOnPostExecute(@NonNull CardBrowser browser, PairWithBoolean<Ca
+    private ExportListener exportListener() {
+        return new ExportListener(this);
+    }
+    private static class ExportListener extends TaskListenerWithContext<CardBrowser, Void, Pair<Boolean, String>>{
```
Yeah, the idea of having a WeakReference of a context and watching it is ingenious idea. I didn't know alot about  WeakReference in java previously

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/libanki/AnkiPackageExporter.java:
```diff
@@ -50,23 +50,35 @@
+        mCids = null;
+    }
+
+    public Exporter(Collection col, Long did, List<Long> mCids) {
```
You are right I meant to add ``public Exporter(Collection col, List<Long> cids) { ... }`` to be consistent with the mDid one. I think in the codebase the setter is used instead of constructor

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -1486,7 +1488,11 @@ public ExportApkg(String apkgPath, Long did, Boolean includeSched, Boolean inclu
                 AnkiPackageExporter exporter = new AnkiPackageExporter(col);
```
Yeah, that is a good idea, should I also remove the setters so it can be immutable-ish ?

Arthur-Milchior replied:
would be perfect.

In this case, since it's a good practice code cleaning that is independant of the feature you had, you may want to consider adding it in a new PR. This new PR would improve code quality, and can be merged really quickly, ensuring that the current PR get smaller and easier to read

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ExportDialog.java:
```diff
@@ -28,8 +37,8 @@
-     * 
-     * @param did An integer which specifies which of the sub-dialogs to show
+     *
+     * @param did           An integer which specifies which of the sub-dialogs to show
```
Yeah, Android studio did that automatically when I added the javadoc comment. I will remove it.

TarekkMA replied:
Maybe a good idea is having ``ankidriod-styles.xml`` for IntelliJ, so it doesn't miss up formatting

Arthur-Milchior replied:
I tried to automate styling in the past but did not find a proper way to do it.
Essentially, anything that do automate styling would make change that conflict with all open pull request, and give a lot of extra work to people just because we took time to merge their pull request. But if you have an idea that would improve this, I'm all ear

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ExportDialog.java:
```diff
@@ -42,6 +51,16 @@ public static ExportDialog newInstance(@NonNull String dialogMessage, Long did)
 
 
+    public static ExportDialog newInstance(@NonNull String dialogMessage, List<Long> cardIds) {
+        ExportDialog f = new ExportDialog();
```
I think the second method is better, I don't like the bundle api.
Exposing it to outside callers will move the hassle of putting values in bundles to them.

Arthur-Milchior replied:
oups, my bad. I meant "private" in both case and not "public". I wanted to keep the two existing methods and have them call the new `newInstance` to avoid duplicate code

TarekkMA replied:
As this is a refactor not related to the code I will push it as a new PR after this one

Arthur-Milchior replied:
Thanks a lot

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ExportDialog.java:
```diff
@@ -51,22 +70,29 @@ public static ExportDialog newInstance(@NonNull String dialogMessage) {
 
+    List<Long> getCardIdsFromArgs() {
+        Serializable cardIdsSer = getArguments().getSerializable("cardIds");
+        return cardIdsSer == null ? null : Arrays.asList((Long[]) cardIdsSer);
```
The bundle API allows ``long[]`` which is primitive type and will a tedious thing to convert ``List<Long>`` to it. So I opted to convert the ``List<Long>`` to ``Long[]`` which does implement the Serializable protocol.

TarekkMA replied:
But I realized now that there is a ``public static long[] collection2Array(java.util.Collection<Long> list)`` in ``Utils.java`` so I will use ``args.putLongArray(...)`` then

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/ExportHelper.java:
```diff
@@ -0,0 +1,210 @@
+
+public class ExportHelper {
+
+    public interface ExportHelperListener extends ExportCompleteDialog.ExportCompleteDialogListener {
```
I thought about using ``ExportCompleteDialog.ExportCompleteDialogListener`` only. But the need may come in the future where ``ExportHelper`` it will need the calling activity to comply with multiple interfaces or have its own methods.

And ``ExportCompleteDialog`` is hidden from the calling activity also. So I thought if the calling activity utilize ``ExportHelper`` so implementing ``ExportHelperListener`` is better

Arthur-Milchior replied:
I'd prefer to make this change if it's required one day, and keep code smaller for know. There is already enough code in the codebase not to add complexity for a potential future.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/ExportHelper.java:
```diff
@@ -0,0 +1,210 @@
+import androidx.core.content.FileProvider;
+import timber.log.Timber;
+
+public class ExportHelper {
```
yeah, I have done this in another commit in the same PR, but I will organize things once I get to split the commits

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:
```diff
@@ -312,11 +315,15 @@ public boolean onLongClick(View v) {
 
     private final ImportAddListener mImportAddListener = new ImportAddListener(this);
+
+
```
Totaly agree

TarekkMA commented:
Thank you about the wonderful feedback. I really appreciate you taking the time to write it.

I will try to implement some tests for the new ported code, will need to study other tests before writing my own.
I'm really sorry about the miss I think android studio did some formatting before I pushed the code. I will try to fix that in its own commit.

I'm not yet familiar with git beyond basic operations, so I will read up on that and try to make more meaningful commits.

About GSoC, yes I contacted you on discord about this PR. I do have some level of experience in android but like you said the change wasn't big, I simply refactored and organized things up. But there is much more I can still learn from being a part of the internship. Being a contributor of a huge open source application like anki is a first for me. And I think anki have alot of things that can be done or learned that I'm not yet aware of. For example the ``TaskListenerWithContext`` with its implementation of ``WeakReference`` is something new, I just learned, and the use of custom web view to handle user content is also new to me.

Again I really appreciate your feedback, and I will try to follow your suggestions. Thank you

TarekkMA commented:
I will not be able to work on this PR for the next few days, as I will be having exams this week. So I will try to work sparingly, or after I finish my exams.

Arthur-Milchior commented:
Please take your time. There is no emergency, AnkiDroid spent 10 years without it, it can wait a month more. We are a tool to help people succeed academically, not make them fail :p

TarekkMA commented on AnkiDroid/src/main/res/values/strings.xml:
```diff
@@ -1,6 +1,8 @@
     <string name="ga_trackingId" translatable="false">UA-125800786-1</string>
+    <string name="card_editor_export_card">Export cards</string>
+    <string name="confirm_apkg_export_selected_cards">Export selected cards as apkg file?</string>
     <!-- If you go over limits, further hits are dropped, setting a sampling percentage can help -->
```
Solved now string in `02-strings.xml` with other string in the appropriate location

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/libanki/AnkiPackageExporter.java:
```diff
@@ -50,23 +50,35 @@
     protected final Collection mCol;
     protected Long mDid;
     protected int mCount;
+    protected List<Long> mCids;
```
Done

TarekkMA replied:
Oh okay ill add it to another commit, was planning to add another PR for code quality improvements so it can be accepted faster.

Arthur-Milchior replied:
That would be perfect. Then the current PR will wait for the other one to go first

TarekkMA commented:
@Arthur-Milchior I have cleaned up the miss, and have divided commits into multiple logical ones, with appropriate messages.

Now there is no unrelated code formatting issues (lines, spaces). I have renamed `ExportHelper` to `ActivityExportingDelegate`.

Very eager to hear your thoughts about the new changes.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/anki/ActivityExportingDelegate.java:
```diff
@@ -0,0 +1,209 @@
+ * <p>
+ * The calling activity must implement {@link ExportCompleteDialogListener} and can then forward any call the the exporting delegate
+ */
+public class ActivityExportingDelegate <A extends AnkiActivity & ExportCompleteDialogListener> {
```
I'm using generics, so I don't have to check if the developer is implementing the interface at runtime, as I have done in the old commits

Arthur-Milchior commented:
Please, if you did a change to some code and that there is no need to answer it, don't hesite to "resolve" the comment. There are a lot of it, and this will win time

Arthur-Milchior commented:
Fun fact: I added to Anki the feature to select card in browser and export them. So this is partially my code that you're porting to java. It's a strange feeling to review it.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:
```diff
@@ -512,6 +535,9 @@ protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         Timber.d("onCreate()");
+
+        mExportingDelegate = new ActivityExportingDelegate(this,PICK_EXPORT_FILE);
```
A space afte the comma, please

Arthur-Milchior reviewed (commented):
I find strange that you write "Individual card" when you take a list of card. It seems like only one card at a time can be exported, while actually you can export as many card as you want thanks to the multi selection in deck.

If it's okay, I'd like to first merge a PR which remove setDid, so that we avoid adding setter and clean the code before comiting it.

Can you please let us know  whether you also tested importing the exported card back ? E.g. by deleting the exported card and importing them again, or sending them to a computer and importing them in anki.

TarekkMA commented on AnkiDroid/src/main/java/com/ichi2/libanki/AnkiPackageExporter.java:
```diff
@@ -311,6 +324,11 @@ public void setIncludeMedia(boolean includeMedia) {
     }
+
+
+    public void setCids(List<Long> cids) {
```
Yeah I will remove all setters as you suggested in another PR

TarekkMA commented:
Thanks a lot @Arthur-Milchior for the feedback.

I'm referring to it as "Individual **cards**" that the user have selected, maybe I'm using the expression wrong, if so should I change the commit messages?

I have tested to export it to another AnkiDriod(Old version) and to my PC, they both seem to work fine (will test it again after the new changes).

Will add nullable annotations and javadocs to the `Exporter`, and add missing formatting to this PR.

Should I remove setters exporter and refactor the `ExportDialog.newInstance` in the same new PR?

Arthur-Milchior commented:
I'm more than fine with having a first PR about cleaning and a second one about the feature. You'll need to rebase current PR over the cleaning one, I'm sorry for that.

I'm not a native english speaker. "individual cards" seems a contradiction to me, but if you believe it's okay for multiple cards, I'm fine with it.

TarekkMA commented:
No issues will push a new PR with that.

 I'm also not a native speaker of English, so I would not know if this is 100% correct, but I feel like if this induces confusion we should rename it to something else? What do you suggest?

Arthur-Milchior commented:
Export selected cards from Cardbrowser

TarekkMA commented:
I have renamed the PR and the previous commit messages.
Will work on the cleaning PR and submit it

mikehardy commented:
This is going to be a very helpful feature, I agree with you @Arthur-Milchior ! Thanks @TarekkMA - and Arthur, when you feel this is good enough I will trust you for sure, you have worked so closely on reviewing it I'm not sure what I would add with limited time available

TarekkMA commented:
@Arthur-Milchior I have rebased this PR over #8196. I think it's ready for a review

TarekkMA commented:
@Arthur-Milchior can you have a look at this.

Arthur-Milchior commented:
There is quite a lot of conflict right now. Would you mind rebasing on master, resolving conflicts. This version clearly won't be merged as is and so I don't want to review it already

TarekkMA commented:
@Arthur-Milchior done.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DeckPickerExportCompleteDialog.java:
```diff
@@ -35,15 +44,15 @@ public MaterialDialog onCreateDialog(Bundle savedInstanceState) {
                 .onPositive((dialog, which) -> {
-                    ((DeckPicker) getActivity()).dismissAllDialogFragments();
-                    ((DeckPicker) getActivity()).emailFile(exportPath);
+                    ((DeckPickerExportCompleteDialogListener) getActivity()).dismissAllDialogFragments();
```
Does the PR with the first commit only compiles and pass tests ? We want all commit to compiles otherwise we can't use git bisect to check for bug.

I believe you must add `DeckPicker extends DeckPickerExportCompleteDialogListener` in the first commit, otherwise casting to `DeckPickerExportCompleteDialogListener` will fail, since `DeckPicker` is not yet a `DeckPickerExportCompleteDialogListener`

TarekkMA replied:
Yes it does implement that.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Utils.java:
```diff
@@ -463,6 +463,21 @@ public static String ids2str(JSONArray ids) {
+    public static List<Long> primitiveArray2List(@NonNull long[] arr) {
+        final List<Long> list = new ArrayList<Long>(arr.length);
+        for (int i = 0; i < arr.length; i++) {
+            list.add(i, arr[i]);
```
Why do you need to indicate the index at which the element is added ? I'd have thought that 
```java
for (Long l: arr) {
  list.add(l);
}
```
would work as well and be simpler

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/ActivityExportingDelegate.java:
```diff
@@ -60,14 +60,19 @@ public void exportApkg(String filename, Long did, boolean includeSched, boolean
             exportPath = new File(exportDir, "All Decks" + timeStampSuffix + ".apkg");
+        } else if (did != null && cardIds == null) {
+            // filename not explicitly specified, but a deck has been specified so use deck name
+            exportPath = new File(exportDir, getCol().getDecks().get(did).getString("name").replaceAll("\\W+", "_") + timeStampSuffix + ".apkg");
```
The commit state "refactored code to reflect such a change" but I don't understand how it is refactoring. It seems that this is a new feature having deck name in file in this function.

By the way, when we export a deck, the file already contains the deck name. I suspect it means that the file name is created somewhere else, it may be better to find where it is done and check whether "Selected cards" should be set there.

TarekkMA replied:
Right, the commit message is not obvious. Will fix that. 

I didn't get what do you mean by the last point, I did mirror the code from collection exporting. Since we can export multiple cards from multiple decks (same as upstream), it doesn't make sense to have the deck name in the file. Also, the upstream file name is "Selected Notes" with not information about the original deck in the file name.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/ActivityExportingDelegate.java:
```diff
@@ -60,14 +60,19 @@ public void exportApkg(String filename, Long did, boolean includeSched, boolean
+            // filename not explicitly specified, but a deck has been specified so use deck name
+            exportPath = new File(exportDir, getCol().getDecks().get(did).getString("name").replaceAll("\\W+", "_") + timeStampSuffix + ".apkg");
+        } else if (did == null && cardIds != null) {
+            exportPath = new File(exportDir, "Selected Cards" + timeStampSuffix + ".apkg");
```
Can you please:
* use a ressource string. So that people get localized file name. 
* Use `Selected_Cards_`. I always feel more confortable if we avoid spaces in filename even when it's not mandatory

TarekkMA replied:
I did hard code it because, other code is doing the same thing. Like in the case of "All Decks"

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:
```diff
@@ -242,6 +247,24 @@
 
+    @Override
+    public void exportApkg(String path, Long did, List<Long> cardIds, boolean includeSched, boolean includeMedia) {
+        mExportingDelegate.exportApkg(path, did, cardIds, includeSched, includeMedia);
```
Unless I'm wrong, it seems duplicate code with DeckPicker.

Couldn't exportApkg and the functions below be defined as [default method](https://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html) in the interface ?

TarekkMA replied:
I could do some experimentation with default methods

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ExportDialog.java:
```diff
@@ -83,8 +83,7 @@ public MaterialDialog onCreateDialog(Bundle savedInstanceState) {
         final Long did = BundleUtils.getNullableLong(getArguments(), "did");
-        final long[] _cids = getArguments().getLongArray("cids");
-        final List<Long> cids = _cids == null ? null : Utils.primitiveArray2List(_cids);
+        final List<Long> cids = BundleUtils.getNullableLongList(getArguments(), "cids");
```
Can you add this in the commit where you added   `final List<Long> cids = _cids == null ? null : Utils.primitiveArray2List(_cids);` ? Seems useless to add an extra commit here

Arthur-Milchior reviewed (commented):
Still a few nits.

As always, I am in favour of splitting PR when they are too big. In this case, I believe that refactoring/extracting to interface could be done in a first PR, and the new feature be added on top of it. This will allow this PR to be simpler to discuss and see first commit merged quicker

TarekkMA commented:
Thanks for the feedback, I will split this into 2 PRs and will also rethink the delegate design since onActivityReuslt will be deprecated soon, and we should move the new API.

Arthur-Milchior commented:
You force pushed. Do you want a review ? Or do i wait for the other pr first ?

TarekkMA commented:
No wa jst rebasing on master and fixing conflicts.

I will split this into 2 PRs as requested.

There is a design chnage I have laid its details on discord. I will open a PR with such a change.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/ActivityExportingDelegate.java:
```diff
@@ -0,0 +1,209 @@
+        // Make sure the file actually exists
+        File attachment = new File(path);
+        if (!attachment.exists()) {
+            Timber.e("Specified apkg file %s does not exist", path);
```
We typically want `Timber.w` for non-fatal errors

TarekkMA replied:
This is from the old code. If the attachment is not found then something did go wrong when exporting the file

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ExportDialog.java:
```diff
@@ -6,16 +6,23 @@
     public interface ExportDialogListener {
 
-        void exportApkg(String path, Long did, boolean includeSched, boolean includeMedia);
+        void exportApkg(String path, Long did, List<Long> cardIds, boolean includeSched, boolean includeMedia);
```
Are there any other designs here? 

Having two mutually exclusive arguments (technically 3 variants - as both can be null) isn't great for understanding.

I'd be tempted to ask if we can express this as a class

TarekkMA replied:
Will redesign the Exporting Delegate so tge activities are completely uncoupled from the dialogs using FragmentFactory (explained on discord).

And in the way I will rethink this one.

david-allison-1 commented on AnkiDroid/src/main/res/values/02-strings.xml:
```diff
@@ -125,6 +125,7 @@
     <string name="card_editor_reposition_card">Reposition</string>
     <string name="card_editor_reset_card">Reset progress</string>
     <string name="card_editor_reschedule_card" comment="Action to reschedule a card">Reschedule</string>
+    <string name="card_editor_export_card" comment="Action to export selected cards">Export cards</string>
```
Just to note: Comments aren't required (but are nice to have) if the string is unique in English

david-allison-1 reviewed (commented):
Love the ActivityExportingDelegate - very clean, an the tests

Few comments RE: design, but I'll let Arthur take the lead on this one

TarekkMA commented:
Keep Open

dae commented:
The UI side of things will likely need updates for the Kotlin conversion. For the backend side of things, the Rust backend supports this use case (https://github.com/ankitects/anki/blob/f6390e9455365f696b0bd06f9be398e6698baf38/proto/anki/import_export.proto#L189), so utilizing it in AnkiDroid with the new schema would just require adding that case to the code (https://github.com/ankidroid/Anki-Android/blob/69da0c89e8795427156bb98afac66414f618a763/AnkiDroid/src/main/java/com/ichi2/libanki/BackendImportExport.kt#L106)

Arthur-Milchior commented:
> The UI side of things will likely need updates for the Kotlin conversion. For the backend side of things, the Rust backend supports this use case (https://github.com/ankitects/anki/blob/f6390e9455365f696b0bd06f9be398e6698baf38/proto/anki/import_export.proto#L189), so utilizing it in AnkiDroid with the new schema would just require adding that case to the code (
> 
> https://github.com/ankidroid/Anki-Android/blob/69da0c89e8795427156bb98afac66414f618a763/AnkiDroid/src/main/java/com/ichi2/libanki/BackendImportExport.kt#L106
> 
> )

Let's also ping @viciousAegis who is already working on the browser

viciousAegis commented:
@dae I am unsure of how to implement the use cases for `cardIds` and `noteIds` mentioned in https://github.com/ankitects/anki/blob/f6390e9455365f696b0bd06f9be398e6698baf38/proto/anki/import_export.proto#L189
This is what I came up with, but it does not seem to be correct.
```kotlin
fun CollectionV16.exportAnkiPackage(
    outPath: String,
    withScheduling: Boolean,
    withMedia: Boolean,
    deckId: DeckId?,
    cardIds: List<CardId>?,
    noteIds: List<NoteId>?,
    legacy: Boolean = true,
) {
    val limit = exportLimit {
        if (deckId != null) {
            this.deckId = deckId
        } else if (cardIds != null) {
            this.cardIds.newBuilderForType().addAllCids(cardIds).build()
        } else if (noteIds != null) {
            this.noteIds.newBuilderForType().addAllNoteIds(noteIds).build()
        } else {
            this.wholeCollection = Empty.getDefaultInstance()
        }
    }
```
Could you guide me in the right direction regarding implementing `cardIds` and `noteIds`? Thanks :)

dae commented:
Try `this.cardIds.cidsList.addAll()`

viciousAegis commented:
> Try `this.cardIds.cidsList.addAll()`

Returns `java.lang.UnsupportedOperationException`, I am assuming because `cidsList` is not mutable.

dae commented:
`this.cardIds = cardIds { cids.addAll() }` ?
